File: src/main/java/lexer/Token.java
package lexer;

public interface Token {

    int tag();

    Token AND = new Symbol('&');
    Token OR = new Symbol('|');
    Token EQ = new Symbol('=');
    Token NOT = new Symbol('!');
    Token LT = new Symbol('<');
    Token GT = new Symbol('>');
    Token PLUS = new Symbol('+');
    Token TIMES = new Symbol('*');
}

File: src/main/kotlin/dragonbook/lexer/Token.kt
package dragonbook.lexer

interface Token {
    val tag: Int

    companion object {
        val AND: Token = Symbol('&')
        val OR: Token = Symbol('|')
        val EQ: Token = Symbol('=')
        val NOT: Token = Symbol('!')
        val LT: Token = Symbol('<')
        val GT: Token = Symbol('>')
        val PLUS: Token = Symbol('+')
        val TIMES: Token = Symbol('*')
        val MINUS: Token = Symbol('-')
        val EOF: Token = Symbol(tag = 65535)
    }
}

data class Symbol(override val tag: Int) : Token {
    constructor(tag: Char) : this(tag.code)

    override fun toString(): String = tag.toChar().toString()
}